buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,403 @@
+    }
+    return charIndex;
+  }
+
+  /**
+   * Gets the function name from the declaration.
+   *
+   * @param functionDeclaration The function declaration of form
+   *                            <return_type> <name> (<args>) :
+   */
+  private String getFunctionName(String functionDeclaration) {
+    Matcher m = namePattern.matcher(functionDeclaration);
+    // Name is the second match.
+    m.find();
+    m.find();
+    return m.group();
+  }
+
+  /**
+   * Responsible for parsing a block of text surrounded by curly braces. It is
+   * assumed that parsing begins at the start of a curly block. Maintains the
+   * state of which structure the parser is currently in to ensure that
+   * encountered curly braces are valid.
+   */
+  static class CurlyParser {
+
+    /**
+     * All of the possible important structures that a token can be inside of.
+     * This is tracked to ensure that when a curly brace is encountered,
+     * whether it is part of the syntax or not can be determined.
+     */
+    enum InsideState {
+
+      /**
+       * Not inside of anything.
+       */
+      NONE,
+
+      /**
+       * Inside of a string declaration (block of double quotes).
+       */
+      STRING,
+
+      /**
+       * Inside of a character declaration (block of single quotes)
+       */
+      CHARACTER,
+
+      /**
+       * Inside of a single line comment.
+       */
+      SINGLE_COMMENT,
+
+      /**
+       * Inside of a multi line comment.
+       */
+      MULTI_COMMENT
+    }
+
+    // Default state is that the parser is not within any structure.
+    private InsideState insideState = InsideState.NONE;
+
+    // Keeps track of the number of open curly braces that have been
+    // legally encountered (those that are not within any structure).
+    private int curlyCounter = 0;
+
+    /**
+     * Parses the given token and updates the state. It is assumed that the
+     * tokens are provided as a sequence:
+     *
+     * for any token1, token2
+     * if parseToken(token1) is called and parseToken(token2) is called right after,
+     * then token2 comes directly after token1 in the stream of tokens.
+     *
+     * @param token The token to parse
+     * @return Whether or not the curly block has been fully parsed
+     */
+    boolean parseToken(String token) {
+      updateState(token);
+      return curlyCounter == 0;
+    }
+
+    /**
+     * Determines the updated state based on the token and current state.
+     *
+     * @param token The token to parse
+     */
+    private void updateState(String token) {
+      if (token.equals("\n")) {
+        if (insideState == InsideState.SINGLE_COMMENT) {
+          insideState = InsideState.NONE;
+        }
+      } else if (token.equals("\"")) {
+        if (insideState == InsideState.NONE) {
+          insideState = InsideState.STRING;
+        } else if (insideState == InsideState.STRING) {
+          insideState = InsideState.NONE;
+        }
+      } else if (token.equals("'")) {
```

Comment:
what about ' tokens in strings? eg " something 'single quotes' something "

Reply:
actually nvm I think the previous case should cover it.

Reply:
Can you add a set of tests in the PR for just this CurlyParser? It's pretty standalone.

Reply:
That is handled already by the state logic, I have a state diagram in my design doc which may help.

Reply:
@AndrewPochapsky and I spoke offline. The plan is to add tests for the entire class in this PR, rather than a follow-up one.